## Машинное обучение 1, ПМИ ФКН ВШЭ

## Практическое домашнее задание 1

### Общая информация

Дата выдачи: 11.09.2020

Мягкий дедлайн: 04:59MSK 21.09.2020 (но лучше сдайте пораньше и успейте выспаться)

Жесткий дедлайн: 23:59MSK 24.09.2020

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В первом разделе вам предстоит выполнить базовые задания с помощью вышеуказанных библиотек, а во втором распределить студентов по курсам. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотек pandas и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов.

Сдавать задание после жёсткого дедлайна нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

### Формат сдачи
Задания сдаются через систему Anytask. Инвайт можно получить у семинариста или ассистента. Присылать необходимо ноутбук с выполненным заданием. Сам ноутбук называйте в формате homework-practice-01-pandas-Username.ipynb, где Username — ваша фамилия.

Для удобства проверки самостоятельно посчитайте свою максимальную оценку (исходя из набора решенных задач) и укажите ниже.

Оценка: xx.

## 0. Введение

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написанием production-кода. 

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*. 

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

In [ ]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

Ячейка с неразмеченыным текстом.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

$F(k) = \int_{-\infty}^{\infty} f(x) e^{2\pi i k} dx$

[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. Выделять *текст* <s>при</s> **необходимости**
2. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

<img src="https://zebra-tv.ru/upload/iblock/cfd/foto_crasycat.jpg" style="width: 100px"> 

А ещё можно вставлять картинки, или гифки, или что захотите:

<img src="https://media1.tenor.com/images/9b8fbe9214504bcf6c60fe4e4f7e114e/tenor.gif?itemid=5416416" style="width: 400px">

### Google Colab

Что за колаб? 

**Google Colab (Colaboratory)** это **Jupyter Notebook + Cloud + Google Drive.**

Компания Google предоставляет возможность бесплатно запускать ноутбуки (предварительно загрузив их на свой гугл-диск) прямо в облаке. При этом вам не требуется установка никаких пакетов на свою машину, а работать можно прямиком из браузера. Вот ссылка:

https://colab.research.google.com

При использовании вы увидете много сходств с jupyter ноутбуком. Одним из преимуществ является доступность GPU, соответствующую опцию можно активировать в настройках сервиса. При желании вы сможете найти в интернете много туториалов по использованию или разобраться самостоятельно =)

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных.

Основной объект в pandas — это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [ ]:
%pylab inline  
import pandas as pd

Populating the interactive namespace from numpy and matplotlib


Скачаем данные:

In [ ]:
!wget  -O 'end_seminar.xlsx' -q 'https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0'

Для пользователей Windows: скачайте файл самостоятельно и поместите его в папку с тетрадкой. Или попробуйте один из следующих вариантов:

In [ ]:
# !powershell iwr -outf somefile https://somesite/somefile

In [ ]:
# !pip install wget
# import wget 
# wget.download('https://dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=1', 'end_seminar.xlsx')

##### В первой части задания (до раздела "Распределение студентов по курсам") использование циклов запрещается и повлечет за собой снижение оценки

Для каждой задачи из этого раздела вы должны написать код для получения ответа, а также дать текстовый ответ, если он предполагается.

На некоторые вопросы вы можете получить путём пристального взгляда на таблицу, но это не будет засчитываться. Вы в любом случае должны получить ответ с помощью кода.

#### 1. [0 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы? (на это не надо отвечать, просто подумайте об этом)

In [ ]:
data = pd.read_excel('end_seminar.xlsx')

In [ ]:
data

timestamp  ... is_ml_student
0   2020-05-15 01:12:50.543  ...           NaN
1   2020-05-15 02:46:48.066  ...           NaN
2   2020-05-15 03:12:41.480  ...           NaN
3   2020-05-15 04:43:08.994  ...           NaN
4   2020-05-15 07:47:17.197  ...           NaN
..                      ...  ...           ...
356 2020-05-19 19:04:17.873  ...           NaN
357 2020-05-19 20:19:25.815  ...           NaN
358 2020-05-20 11:54:38.318  ...           NaN
359 2020-05-23 21:16:40.916  ...           NaN
360 2020-05-23 21:16:40.916  ...           NaN

[361 rows x 16 columns]

In [ ]:
data.tail(10)

timestamp  ... is_ml_student
351 2020-05-18 23:58:33.903  ...           NaN
352 2020-05-18 23:59:20.831  ...           NaN
353 2020-05-18 23:59:28.998  ...           NaN
354 2020-05-19 08:55:39.509  ...           NaN
355 2020-05-19 18:58:06.283  ...           NaN
356 2020-05-19 19:04:17.873  ...           NaN
357 2020-05-19 20:19:25.815  ...           NaN
358 2020-05-20 11:54:38.318  ...           NaN
359 2020-05-23 21:16:40.916  ...           NaN
360 2020-05-23 21:16:40.916  ...           NaN

[10 rows x 16 columns]

In [ ]:
data.columns

Index(['timestamp', 'id', 'rating', '18_group', 'is_mi', 'fall_1', 'fall_2',
       'fall_3', 'spring_1', 'spring_2', 'spring_3', 'is_first_time',
       'percentile', '17_group', 'blended', 'is_ml_student'],
      dtype='object')

#### 2. [0.5 балла] Ответьте на вопросы:
1. Сколько было заявок из групп 18-го года набора, а сколько из групп 17-го года?
2. Есть ли студенты с равными перцентилями?

In [ ]:
len(data[data['18_group'].isin(['NaN'])])

138

заявок из групп 17-го года: 138

In [ ]:
len(data[data['17_group'].isin(['NaN'])])

223

заявок из групп 18-го года: 223

1. С помощью метода .isin выведем лишь те строки нашей таблицы, в которых в колонке 18_group имеем значения NaN, то есть будут выведены студенты 17-ого года , которые подавали заявки. Их будет 138. А после выведем аналогично для столбца  17_group и получим кол-во заявок 18-ого года соотв. Их - 223. Также замечаем, что нет такого, что у студента в обоих признаках 18_group и  17_group стоит NaN и в сумме кол-во заявок равно кол-ву всех студентов в таблице 

In [ ]:
len(data['percentile'].unique())

346

2. Посчитав кол-во уникальных значений столбца 'percentile' замечаем, что 346 < 361, следовательно в данных существуют одинаковые перцентили.

#### 3. [0.5 балла] Есть ли в данных пропуски? В каких колонках? Сколько их в каждой из этих колонок?

In [ ]:
data.isna().sum()

timestamp          0
id                 0
rating             0
18_group         138
is_mi            343
fall_1             0
fall_2             0
fall_3             0
spring_1           0
spring_2           0
spring_3           0
is_first_time      2
percentile         0
17_group         223
blended          223
is_ml_student    304
dtype: int64

В данных существуют пропуски в колонках: \\
18_group    -     138 \\
is_mi        -    343 \\
is_first_time -     2 \\
17_group       -  223 \\
blended         - 223 \\
is_ml_student    -304 \\




Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [ ]:
data.dtypes

timestamp        datetime64[ns]
id                       object
rating                    int64
18_group                float64
is_mi                   float64
fall_1                   object
fall_2                   object
fall_3                   object
spring_1                 object
spring_2                 object
spring_3                 object
is_first_time            object
percentile              float64
17_group                float64
blended                  object
is_ml_student           float64
dtype: object

In [ ]:
data['18_group'] = data['18_group'].fillna(value= 0)

In [ ]:
data['is_mi'] = data['is_mi'].fillna(value= 0)

In [ ]:
data['17_group'] = data['17_group'].fillna(value= 0)

In [ ]:
data['is_ml_student'] = data['is_ml_student'].fillna(value= 0)

In [ ]:
data['is_first_time'] = data['is_first_time'].fillna(value= '')

In [ ]:
data['blended'] = data['blended'].fillna(value= '')

In [ ]:
data.isna().sum()

timestamp        0
id               0
rating           0
18_group         0
is_mi            0
fall_1           0
fall_2           0
fall_3           0
spring_1         0
spring_2         0
spring_3         0
is_first_time    0
percentile       0
17_group         0
blended          0
is_ml_student    0
dtype: int64

Проверили, пропусков больше нет!

#### 4. [0.5 балла] Посмотрите повнимательнее на колонку 'is_first_time'. 

Есть ли в ней ответы "Нет"? Сколько их?

Если вы найдете повторные обращения студентов, оставьте только самую позднюю версию. 

In [ ]:
data['is_first_time'].value_counts()

Да     308
Нет     51
         2
Name: is_first_time, dtype: int64

Взяла код [отсюда](https://stackoverflow.com/questions/22391433/count-the-frequency-that-a-value-occurs-in-a-dataframe-column). Теперь уберем повторные обращения:

In [ ]:
data['id'].value_counts()

ffc58d371e2f52e1c7f45d49d3d629d0    2
86c24daaa62246142d46ee58534a65e0    2
b121e4d7d62387cd5199cca464a34a6d    2
47e2d8bbb630bc2d9b3fbf98fe77cffb    2
5c49473012411a1eee72f8ea63767adf    2
                                   ..
a207faa550aba9ebf30537be445012cc    1
b3cbd3541c0cf66ff4eb2bc1596bb149    1
b8dab0479a00af2eb75deaf54dd9a442    1
9fa4bc4eb50e574e88b670a86781860b    1
01b072a0cdb45814fc4ab8ca3e621f0e    1
Name: id, Length: 347, dtype: int64

In [ ]:
data['id_count'] = data['id'].map(data['id'].value_counts())
data = data.sort_values(by=['timestamp'])


In [ ]:
data[data['id_count'] == 2]

timestamp  ... id_count
22  2020-05-15 22:15:31.902  ...        2
23  2020-05-15 22:30:31.902  ...        2
29  2020-05-16 14:46:06.274  ...        2
30  2020-05-16 15:01:06.274  ...        2
257 2020-05-17 15:30:40.828  ...        2
259 2020-05-17 15:55:40.828  ...        2
97  2020-05-18 18:38:41.094  ...        2
98  2020-05-18 18:53:41.094  ...        2
300 2020-05-18 19:58:36.111  ...        2
302 2020-05-18 20:23:36.111  ...        2
121 2020-05-18 20:47:28.759  ...        2
126 2020-05-18 21:02:28.759  ...        2
129 2020-05-18 21:12:02.445  ...        2
139 2020-05-18 21:27:02.445  ...        2
141 2020-05-18 21:34:22.233  ...        2
144 2020-05-18 21:36:52.845  ...        2
150 2020-05-18 21:49:22.233  ...        2
152 2020-05-18 21:51:52.845  ...        2
321 2020-05-18 22:15:28.531  ...        2
174 2020-05-18 22:36:34.323  ...        2
325 2020-05-18 22:40:28.531  ...        2
176 2020-05-18 22:51:34.323  ...        2
333 2020-05-18 22:59:48.715  ...        2
186 2020-05-18 23:20:28.235  ...        2
342 2020-05-18 23:24:48.715  ...        2
192 2020-05-18 23:35:28.235  ...        2
196 2020-05-18 23:40:41.196  ...        2
204 2020-05-18 23:55:41.196  ...        2

[28 rows x 17 columns]

In [ ]:
len(data[data['id_count'] == 2])

28

In [ ]:
data.drop_duplicates(subset=['id'], keep = 'last', inplace = True) #так как мы отсортировали по времени наши строки, то с помощью аргумента keep, мы оставим только последнюю из всех версий по одинаковому id.

Проверим, что все получилось: \\
Действительно мы ниже получаем, что удалилось ровно 28:2 = 14 строк.

In [ ]:
data['id'].value_counts()

c976c868dd660e792ab82ba8d9e02506    1
2ba0a356c95af4fe74272e708082086f    1
425c2a90f447620c9a1868e92f0a7af6    1
cc8e710c871ae8236b8b8705e6840793    1
851497aadac83e844753bb3b2163c79b    1
                                   ..
9fa4bc4eb50e574e88b670a86781860b    1
a69948a1730e412eab71a4d4e70791e2    1
e0fceae1e285f737630bb6123a1bd85d    1
31e38f7596f83b72ebba1c6fc6b7a590    1
01b072a0cdb45814fc4ab8ca3e621f0e    1
Name: id, Length: 347, dtype: int64

In [ ]:
data[data['id_count'] == 2]

timestamp  ... id_count
23  2020-05-15 22:30:31.902  ...        2
30  2020-05-16 15:01:06.274  ...        2
259 2020-05-17 15:55:40.828  ...        2
98  2020-05-18 18:53:41.094  ...        2
302 2020-05-18 20:23:36.111  ...        2
126 2020-05-18 21:02:28.759  ...        2
139 2020-05-18 21:27:02.445  ...        2
150 2020-05-18 21:49:22.233  ...        2
152 2020-05-18 21:51:52.845  ...        2
325 2020-05-18 22:40:28.531  ...        2
176 2020-05-18 22:51:34.323  ...        2
342 2020-05-18 23:24:48.715  ...        2
192 2020-05-18 23:35:28.235  ...        2
204 2020-05-18 23:55:41.196  ...        2

[14 rows x 17 columns]

#### 5. [0.5 балла] Какие  blended-курсы для четверокурсников существуют? На какой blended-курс записалось наибольшее количество студентов? На каком из курсов собрались студенты с самым высоким средним рейтингом? 

In [ ]:
data[data['18_group'] == 0]['blended'].value_counts()

DevOps                                   87
Соревновательный анализ данных           32
Введение в дифференциальную геометрию    15
Name: blended, dtype: int64

In [ ]:
max(data[data['18_group'] == 0]['blended'].value_counts())

87

Существуют blended курсы: \\
DevOps                              
Соревновательный анализ данных          
Введение в дифференциальную геометрию \\
Больше всего записалось на курс DevOps, таких 87 студентов.

In [ ]:
data[data['18_group'] == 0].groupby(by='blended').mean()['rating']

blended
DevOps                                   1154.172414
Введение в дифференциальную геометрию    1210.733333
Соревновательный анализ данных           1144.218750
Name: rating, dtype: float64

In [ ]:
data[data['18_group'] == 0].groupby(by='blended').mean()['rating'].max()

1210.7333333333333

Из выведенных данных можно заметить, что максимальное среднее по рейтингу будет на курсе **Введение в дифференциальную геометрию**. 

#### 6. [1 балл] Выясните, есть ли в данных студенты с абсолютно одинаковыми предпочтениями по всем курсам (не забудьте учесть blended-курсы для четверокурсников). Сколько таких наборов, которые взяли несколько студентов? Выведите их все вместе с количеством выбравших их студентов.

In [ ]:
data_dublicates = data.duplicated(subset=['fall_1', 'fall_2',
       'fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'], keep=False)

len(data[data_dublicates])

32

In [ ]:
data[data_dublicates].groupby(['fall_1', 'fall_2','fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'],as_index=False).size()


fall_1                                           fall_2                                           fall_3                                           spring_1                                                spring_2                                     spring_3                                     blended
Statistical Learning Theory                      Анализ неструктурированных данных                Сбор и обработка данных с помощью краудсорсинга  Обучение с подкреплением                                Конфликты и кооперация                       Компьютерное зрение                          DevOps     2
Байесовские методы машинного обучения            Анализ неструктурированных данных                Глубинное обучение в обработке звука             Обучение с подкреплением                                Генеративные модели в машинном обучении      Компьютерное зрение                          DevOps     2
                                                                                       

In [ ]:
len(data[data_dublicates].groupby(['fall_1', 'fall_2','fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'],as_index=False).size())

15

15 различных наборов, которые выбрали несколько человек

#### 7. [0.5 балла] Найдите курсы по выбору, на которые записывались как студенты 18-го года набора, так и студенты 17-го года.

In [ ]:
# your code

Методом исключения найдите курсы, которые предлагались только студентам 18-го года и только студентам 17-го года.

In [ ]:
# your code

### Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. Здесь будут описаны ключевые правила оформления графиков для **всех** домашних заданий.

У matplotlib, конечно же, есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторой величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить один из них. Не забывайте про базовые принципы построения приличных графиков:
- оси должны быть подписаны, причём не слишком мелко;
- у графика должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах);
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать;
- если распределение на гистограмме имеет тяжёлые хвосты, лучше использовать логарифмическую шкалу.

In [ ]:
%matplotlib inline
import matplotlib.pyplot as plt

#### 8. [0.5 балла] Постройте график средних рейтингов по дням получения ответов (bar plot).

In [ ]:
# your code

Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
# your code

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё. 

### 2. Распределение студентов по курсам.

Теперь вам нужно распределить студентов по осенним курсам по выбору, учитывая их предпочтения.

Алгоритм распределения студентов по курсам:
1. По умолчанию на каждой дисциплине по выбору у 3 и 4 курсов может учиться 1 группа (до 30 студентов). Исключения описаны ниже. На blended-дисциплинах для четверокурсников количество мест не ограничено.
2. Проводится первая волна отбора. Для каждой дисциплины формируется список тех, кто указал её первым приоритетом (если студент должен выбрать два курса по выбору, то для него дисциплины, которые он указал первым и вторым приоритетом, рассматриваются как дисциплины первого приоритета). Если желающих больше, чем мест, то выбирается топ по перцентилю рейтинга.
3. На дисциплинах, где остались места после первой волны, формируются списки тех, кто выбрал их вторым приоритетом, и места заполняются лучшими по перцентили рейтинга студентами. После этого проводится такая же процедура для дисциплин третьего приоритета.
4. Если студент не попал на необходимое количество курсов по итогам трёх волн, с ним связывается учебный офис и решает вопрос в индивидуальном порядке.

Обращаем ваше внимание на следующие детали:

- По умолчанию студент выбирает один осенний и один весенний курс по выбору, а также четверокурсники выбирают один blended-курс. Студенты групп 17-го года специализаций МОП и ТИ выбирают по 2 осенних и 2 весенних курса по выбору, также студенты групп 18' специализации МИ выбирают 2 осенних курса.

- Студенты специализации МОП не могут выбрать весенним курсом по выбору Машинное обучение 2.

- Blended-курсы не трогайте, по ним не надо распределять, на другие курсы они никак не влияют.

- Постарайтесь воздержаться от использования циклов там, где это возможно.

- На выходе ожидается файл res_fall.csv с результатами распределения на осенние курсы по выбору. Файл должен быть следующего формата:

    * три столбца: ID, course1, course2
    
    * Если студент не попал на курс, но должен был, то вместо названия курса в ячейке должна быть строка "???"
    
    * Если студент должен выбрать только один курс, то в колонке course2 для него должна стоять строка "-"
    
    * Формат csv: для сохранения воспользуйтесь df.to_csv('solution.csv', index=None)
    

Для работы вам могут понадобиться следующие данные:

- Результаты опроса (вы уже использовали этот файл в первой части задания, но на всякий случай ссылка: https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0)

- Соответствие номеров групп специализациям:

    * 171, 172 - МОП; 173 - ТИ; 174 — АДИС; 175, 176 — РС; 177 — АПР
    
    * 181, 182 - МОП; 183 - ТИ; 184 - АДИС; 185, 186 - РС; 187 - АПР; 188 - МИ 

- Ограничения по количеству мест на курсах по выбору:

    * Осенние: везде 30 мест, кроме Statistical Learning Theory (60 мест), Высокопроизводительных вычислений (60 мест), Анализа неструктурированных данных ($\infty$ мест)

    * Весенние: везде 30 мест, кроме Обучения с подкреплением (60 мест), Анализа данных в бизнесе (60 мест).


Кстати, убедитесь, что в данных больше нет пропусков и повторных записей.

#### 0. Проверка

Для начала давайте убедимся, что вы успешно выполнили задания первой части и проверим ваши данные на наличие пропусков и повторов:

In [ ]:
assert data.shape[0] == 347, 'В таблице остались повторы или потеряны данные'

assert data.isna().sum().sum() == 0, 'В таблице остались пропуски'

Если вы не получили AssertionError, то можете продолжать.

#### 1. [1 балл] Создайте новый признак, обозначающий, сколько осенних курсов должен выбрать студент

В этом вам может помочь информация о специализации и группе стундента.

In [ ]:
data['fall_courses'] = np.ones(shape=(data.shape[0],1), dtype=numpy.int8)

In [ ]:
data[(data['18_group'] == 0) & (data['is_mi'] == 1)].shape

(0, 18)

Проверили, что нет таких студентов из 17-ого, которые также являются 'is_mi'.

In [ ]:
data['fall_courses'][data['17_group'] == 171] = data['fall_courses'][data['17_group'] == 171 ].replace(1,2)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
data['fall_courses'][data['17_group'] == 173] = data['fall_courses'][data['17_group'] == 173 ].replace(1,2)

In [ ]:
data['fall_courses'][data['is_mi'] == 1] = data['fall_courses'][data['is_mi'] == 1 ].replace(1,2)

In [ ]:
data['fall_courses'][data['17_group'] == 172] = data['fall_courses'][data['17_group'] == 172 ].replace(1,2)

In [ ]:
len(data[data['fall_courses'] ==2])

89

Проверка:

In [ ]:
col_name = 'fall_courses'   # insert your new column name as str

assert(data[data['id'] == '2662600c2c37e11e62f6ee0b88452f22'][col_name] == 2).all()
assert(data[data['id'] == 'd555d2805e1d93d4f023e57dc4c8f403'][col_name] == 2).all()
assert(data[data['id'] == '8fe79f84f36e3a5d2d6745621321302c'][col_name] == 1).all()
assert(data[data['id'] == 'e4caca755ee0bdd711e18fb8084958b5'][col_name] == 1).all()

#### 2. [2 балла] Распределите студентов в соответствии с первым приоритетом

In [ ]:
courses_fall = (data['fall_1'].append(data['fall_2']).append(data['fall_3'])).unique()

data['course1'] = ''
data['course2'] = ''

dict_of_place_for_courses = {}
dict_of_place_for_courses = dict_of_place_for_courses.fromkeys(courses_fall,0)

data = data.sort_values(by='percentile')

id = list(data['id'].unique())

Взяла для первой строки тут [Код](https://thispointer.com/pandas-get-unique-values-in-single-or-multiple-columns-of-a-dataframe-in-python/#:~:text=To%20get%20the%20unique%20values,on%20that%20series%20object%20i.e.&text=It%20returns%20the%20count%20of%20unique%20elements%20in%20multiple%20columns.)

Проверим, есть ли такие, которые поставили на первый и второй приоритет одно и тоже...:

In [ ]:
data[data['fall_1'] == data['fall_2']] 

timestamp                                id  ...  course1  course2
19  2020-05-15 21:34:32.550  86d929f6fd741e56e93cee5900298849  ...                  
239 2020-05-14 23:33:21.310  e0fceae1e285f737630bb6123a1bd85d  ...                  
283 2020-05-18 14:55:38.053  0d85933f07516e8871e9415c1d95ddbe  ...                  
295 2020-05-18 18:27:32.822  7fcf56aae6836855a689000a50e80360  ...                  

[4 rows x 20 columns]

In [ ]:
dict_of_number_of_places = {}
dict_of_number_of_places = dict_of_number_of_places.fromkeys(courses_fall,30)
dict_of_number_of_places['Statistical Learning Theory'] = 60
dict_of_number_of_places['Высокопроизводительные вычисления'] = 60
dict_of_number_of_places['Анализ неструктурированных данных'] = 999
dict_of_number_of_places

{'Statistical Learning Theory': 60,
 'Анализ неструктурированных данных': 999,
 'Байесовские методы машинного обучения': 30,
 'Безопасность компьютерных систем': 30,
 'Высокопроизводительные вычисления': 60,
 'Глубинное обучение в обработке звука': 30,
 'Матричные вычисления': 30,
 'Моделирование временных рядов': 30,
 'Принятие решений в условиях риска и неопределённости': 30,
 'Сбор и обработка данных с помощью краудсорсинга': 30,
 'Теория баз данных': 30,
 'Язык SQL': 30}

In [ ]:
for i in id:
  fall_1 = data[data['id'] == i]['fall_1'].tolist()[0]
  fall_1_courses = data[data['id'] == i]['fall_courses'].tolist()[0]

  if fall_1_courses == 1:
    if dict_of_place_for_courses[fall_1] < dict_of_number_of_places[fall_1]:
      data.loc[data.id == i, 'course1'] = fall_1
      data.loc[data.id == i, 'course2'] = '-'
      dict_of_place_for_courses[fall_1] += 1
      


  if fall_1_courses == 2 :
    
    fall_2 = data[data['id'] == i]['fall_2'].tolist()[0]
    if fall_1 != fall_2:
    
      if dict_of_place_for_courses[fall_1] < dict_of_number_of_places[fall_1]:
        data.loc[data.id == i, 'course1'] = fall_1
        dict_of_place_for_courses[fall_1] +=1
    
      elif dict_of_place_for_courses[fall_2] < dict_of_number_of_places[fall_2] and (fall_1 != fall_2):
          data.loc[data.id == i, 'course2'] = fall_2
          dict_of_place_for_courses[fall_2] +=1
        



In [ ]:
dict_of_place_for_courses

{'Statistical Learning Theory': 26,
 'Анализ неструктурированных данных': 34,
 'Байесовские методы машинного обучения': 17,
 'Безопасность компьютерных систем': 30,
 'Высокопроизводительные вычисления': 40,
 'Глубинное обучение в обработке звука': 12,
 'Матричные вычисления': 14,
 'Моделирование временных рядов': 23,
 'Принятие решений в условиях риска и неопределённости': 6,
 'Сбор и обработка данных с помощью краудсорсинга': 30,
 'Теория баз данных': 11,
 'Язык SQL': 28}

Здесь для проверки приведена таблица, в которой есть 2 дополнительные колонки:
    
    1) is_first_place - является ли студент лучшим по перцентили хотя бы на одном из курсов, куда он был зачислен 
    (True / NaN)
    
    2) is_last_place  - является ли студент худшим по перцентили хотя бы на одном из курсов, куда он был зачислен (True / NaN)

In [ ]:
!wget  -O '2_task_check.csv' -q 'https://www.dropbox.com/s/v8o2zzq3iz5gc9w/_2_task_check.csv?dl=0'
check_df = pd.read_csv('2_task_check.csv')

После распределения студентов в соответствии с первым приоритетом добавьте в свой датафрейм аналогичные признаки и запустите проверку:

In [ ]:
fir_col_name = 'is_first_place'      # insert name of your new column with is_first_place as str
last_col_name = 'is_last_place'     # insert name of your new column with is_last_place as str


assert((data[data[fir_col_name].isna() == False][['id']].sort_values('id').reset_index(drop=True)
        ==
        check_df[check_df['is_first_place'].isna() == False][['id']].sort_values('id').reset_index(drop=True)
       ).id.values).all()


assert((data[data[last_col_name].isna() == False][['id']].sort_values('id').reset_index(drop=True)
       == 
       check_df[check_df['is_last_place'].isna() == False][['id']].sort_values('id').reset_index(drop=True)
      ).id.values).all()

#### 3. [3 балла] Проведите все три волны отбора студентов на курсы по выбору

In [ ]:
for i in id:
  fall_1 = data[data['id'] == i]['fall_1'].tolist()[0]
  fall_2 = data[data['id'] == i]['fall_2'].tolist()[0]
  fall_3 = data[data['id'] == i]['fall_3'].tolist()[0]
  fall_1_courses = data[data['id'] == i]['fall_courses'].tolist()[0]

  if fall_1_courses == 1 and (fall_1 != fall_2) and (data[data['id'] == i]['course1'] == '').tolist()[0]:
    if dict_of_place_for_courses[fall_2] < dict_of_number_of_places[fall_2]:
      data.loc[data.id == i, 'course1'] = fall_2
      data.loc[data.id == i, 'course2'] = '-'
      dict_of_place_for_courses[fall_2] += 1
      


  if fall_1_courses == 2 and (fall_1 != fall_2 != fall_3):
    
    
    
    if dict_of_place_for_courses[fall_3] < dict_of_number_of_places[fall_3]:
      if (data[data['id'] == i]['course1'] == '').tolist()[0]:
        
        data.loc[data.id == i, 'course1'] = fall_3
        dict_of_place_for_courses[fall_3] +=1
      if (data[data['id'] == i]['course2'] == '').tolist()[0]:
        
        data.loc[data.id == i, 'course2'] = fall_3
        dict_of_place_for_courses[fall_3] +=1
        
      
    
    

In [ ]:
dict_of_place_for_courses

{'Statistical Learning Theory': 37,
 'Анализ неструктурированных данных': 45,
 'Байесовские методы машинного обучения': 26,
 'Безопасность компьютерных систем': 30,
 'Высокопроизводительные вычисления': 60,
 'Глубинное обучение в обработке звука': 29,
 'Матричные вычисления': 23,
 'Моделирование временных рядов': 30,
 'Принятие решений в условиях риска и неопределённости': 10,
 'Сбор и обработка данных с помощью краудсорсинга': 30,
 'Теория баз данных': 17,
 'Язык SQL': 30}

In [ ]:
for i in id:
  fall_1 = data[data['id'] == i]['fall_1'].tolist()[0]
  fall_2 = data[data['id'] == i]['fall_2'].tolist()[0]
  fall_3 = data[data['id'] == i]['fall_3'].tolist()[0]
  fall_1_courses = data[data['id'] == i]['fall_courses'].tolist()[0]

  if fall_1_courses == 1 and (fall_1 != fall_2 != fall_3) and (data[data['id'] == i]['course1'] == '').tolist()[0]:
    if dict_of_place_for_courses[fall_3] < dict_of_number_of_places[fall_3]:
      data.loc[data.id == i, 'course1'] = fall_3
      data.loc[data.id == i, 'course2'] = '-'
      dict_of_place_for_courses[fall_3] += 1

  

In [ ]:
dict_of_place_for_courses

{'Statistical Learning Theory': 38,
 'Анализ неструктурированных данных': 46,
 'Байесовские методы машинного обучения': 26,
 'Безопасность компьютерных систем': 30,
 'Высокопроизводительные вычисления': 60,
 'Глубинное обучение в обработке звука': 29,
 'Матричные вычисления': 30,
 'Моделирование временных рядов': 30,
 'Принятие решений в условиях риска и неопределённости': 11,
 'Сбор и обработка данных с помощью краудсорсинга': 30,
 'Теория баз данных': 21,
 'Язык SQL': 30}

In [ ]:
data

timestamp  ...                                            course2
80  2020-05-18 17:09:23.589  ...                                                  -
231 2020-05-14 21:52:43.844  ...                        Statistical Learning Theory
91  2020-05-18 18:03:57.347  ...                                                  -
273 2020-05-18 12:39:47.722  ...                                                  -
19  2020-05-15 21:34:32.550  ...                                                  -
..                      ...  ...                                                ...
359 2020-05-23 21:16:40.916  ...  Принятие решений в условиях риска и неопределё...
170 2020-05-18 22:29:14.162  ...                                                   
128 2020-05-18 21:12:00.134  ...                                                   
340 2020-05-18 23:12:29.049  ...                                                  -
217 2020-05-20 11:34:10.787  ...                                                  -

[347 rows x 20 columns]

In [ ]:
data.loc[data.course1 == '', 'course1' ] = '???'

In [ ]:
data.loc[data.course2 == '', 'course2' ] = '???'

In [ ]:
data_courses = data[['id','course1', 'course2']]

In [ ]:
data_courses.to_csv('res_fall.csv', index=None)


**Отправьте свой файл res_fall.csv в контест (https://contest.yandex.ru/contest/19782/problems/A/) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

*Дисклеймер:*

Контест выдаётся для самопроверки. Если ваша посылка получила ОК, то код, скорее всего, правильный. Но при этом оценка всё равно может быть снижена в случае обнаружения неэффективностей или ошибок в коде. Если вы сдадите в AnyTask очевидно неработающий код или ноутбук без кода, но при этом в контест будет сдан корректный файл, то это будет расцениваться как плагиат.

На всякий случай просим вас сдать вместе с ноутбуком файл res_fall.csv в anytask

**Дополнительное задание. [2 бонусных балла] Распределите таким же образом студентов еще и на весенние курсы по выбору.**

Если ваш код был хорошо структурирован, то это не составит проблем. 

Если вы выполнили это задание, сдайте среди прочего файл res_spring.csv в таком же формате, как и res_fall.csv.

In [ ]:
# (￣^'￣)ゞ

**Отправьте свой файл res_spring.csv в контест (https://contest.yandex.ru/contest/19782/problems/B/) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

На всякий случай просим вас сдать вместе с ноутбуком файл res_spring.csv в anytask

Вставьте картинку, описывающую ваш опыт выполнения этого задания: